File: apps/android/app/src/main/java/com/velix/app/core/Instance.kt
package com.velix.app.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

@Serializable
data class Instance(
    val id: String,
    val baseUrl: String,
    val displayName: String,
    val userEmail: String,
    val accessToken: String,
)

/** Lista de instâncias Velix logadas — "multi-servidor" no app é
 * multi-instância, não multi-servidor dentro de uma instância só (ver spec de
 * UX, seção 2). */
class InstanceStore(private val store: KeyValueStore) {
    private val json = Json { ignoreUnknownKeys = true }
    private val storageKey = "instances"
    private val activeKey = "active_instance_id"

    private val _instances = MutableStateFlow<List<Instance>>(emptyList())
    val instances: StateFlow<List<Instance>> = _instances.asStateFlow()

    private val _activeInstance = MutableStateFlow<Instance?>(null)
    val activeInstance: StateFlow<Instance?> = _activeInstance.asStateFlow()

    init { load() }

    fun add(instance: Instance) {
        _instances.update { it + instance }
        if (_activeInstance.value == null) _activeInstance.value = instance
        persist()
    }

    fun remove(instance: Instance) {
        _instances.update { list -> list.filterNot { it.id == instance.id } }
        if (_activeInstance.value?.id == instance.id) {
            _activeInstance.value = _instances.value.firstOrNull()
        }
        persist()
    }

    fun setActive(instance: Instance) {
        if (_instances.value.none { it.id == instance.id }) return
        _activeInstance.value = instance
        persist()
    }

    private fun persist() {
        store.save(storageKey, json.encodeToString(_instances.value))
        _activeInstance.value?.let { store.save(activeKey, it.id) } ?: store.delete(activeKey)
    }

    private fun load() {
        val raw = store.read(storageKey) ?: return
        val decoded = try { json.decodeFromString<List<Instance>>(raw) } catch (e: Exception) { return }
        _instances.value = decoded
        val activeId = store.read(activeKey)
        _activeInstance.value = decoded.firstOrNull { it.id == activeId } ?: decoded.firstOrNull()
    }
}

File: apps/ios/Velix/Core/Instance.swift
import Foundation
import Observation

struct Instance: Codable, Identifiable, Equatable {
    let id: UUID
    let baseURL: URL
    let displayName: String
    let userEmail: String
    var accessToken: String
}

extension Instance {
    /// Monta a `Instance` a partir da resposta de `/auth/login` (usada tanto no
    /// login direto quanto após confirmar o 2FA — ver Onboarding/LoginView e
    /// Onboarding/TwoFactorView). `nil` se a resposta não trouxer token/usuário.
    init?(baseURL: URL, loginResponse: LoginResponse) {
        guard let token = loginResponse.accessToken, let user = loginResponse.user else { return nil }
        self.init(
            id: UUID(),
            baseURL: baseURL,
            displayName: baseURL.host ?? baseURL.absoluteString,
            userEmail: user.email,
            accessToken: token
        )
    }
}

/// Lista de instâncias Velix logadas (ver spec de UX, seção 2 — "multi-servidor"
/// no app é multi-instância, não multi-servidor dentro de uma instância só).
@Observable
final class InstanceStore {
    private let keychain: KeychainStore
    private let storageKey = "instances"
    private let activeKey = "active-instance-id"

    private(set) var instances: [Instance] = []
    private(set) var activeInstance: Instance?

    init(keychain: KeychainStore) {
        self.keychain = keychain
        load()
    }

    func add(_ instance: Instance) {
        instances.append(instance)
        if activeInstance == nil {
            activeInstance = instance
        }
        persist()
    }

    func remove(_ instance: Instance) {
        instances.removeAll { $0.id == instance.id }
        if activeInstance?.id == instance.id {
            activeInstance = instances.first
        }
        persist()
    }

    func setActive(_ instance: Instance) {
        guard instances.contains(where: { $0.id == instance.id }) else { return }
        activeInstance = instance
        persist()
    }

    func clearAllForTesting() {
        instances = []
        activeInstance = nil
        persist()
    }

    private func persist() {
        if let data = try? JSONEncoder().encode(instances) {
            keychain.save(data, forKey: storageKey)
        }
        if let activeInstance, let idData = activeInstance.id.uuidString.data(using: .utf8) {
            keychain.save(idData, forKey: activeKey)
        } else {
            keychain.delete(forKey: activeKey)
        }
    }

    private func load() {
        guard let data = keychain.read(forKey: storageKey),
              let decoded = try? JSONDecoder().decode([Instance].self, from: data) else { return }
        instances = decoded
        if let idData = keychain.read(forKey: activeKey),
           let idString = String(data: idData, encoding: .utf8),
           let id = UUID(uuidString: idString) {
            activeInstance = instances.first { $0.id == id }
        } else {
            activeInstance = instances.first
        }
    }
}
